### DISCUSSION TITLE
Better support for AlpineJS

### DISCUSSION THREAD

**Discussion by JuroOravec** [django-components/django-components#471] [Ideas]
## Description

In my project, we use Django + AlpineJS, I've made a few changes that allow me to pass AlpineJS bindings through components more intuitively. Basically it's the same concept as [Vue's passthrough/fallthrough attributes](https://vuejs.org/guide/components/attrs#fallthrough-attributes).

So I can pass AlpineJS syntax like `@click.stop` to my Django components:

```django
{% component "icon"
  icon="outline_trash"
  @click.stop="$dispatch('user_delete', { role })"
%}{% endcomponent %}
```

Whereas the `icon` component looks roughly something like this:

```django
<div {{ attrs }}>
    <span>
        <svg ...>
        {% for path in icon_paths %}
          <path stroke-linecap="round" stroke-linejoin="round" d="{{ path }}" />
        {% endfor %}
        </svg>

        {% slot "content" default %}{% endslot %}
    </span>
</div>
```

3 things happen behind the scene in the example above:
1. I can pass component kwargs with special characters like `@`, `.`, `-`. So `@click.stop` is a valid key.
2. In my `icon` component, I check the names of all args and kwargs of the `get_context_data()` method. All UNKNOWN kwargs get collected into an `attrs` variable.
3. I assign the `attrs` variable to the top level `<div>`, so
	```django
	<div {{ attrs }}>
	...
	```
	ends up rendered as
	```html
	<div @click.stop="$dispatch('user_delete', { role })">
	...
	```

What do you think?

## Notes
1. For allowing extra special characters in kwargs, the inspiration comes from [django-slippers](https://github.com/mixxorz/slippers), which allows `@.:-` in kwargs ([see implementation](https://github.com/mixxorz/slippers/blob/main/slippers/template.py)). I suggest to add also `#` to the mix.
2. With recent introduction of `component_vars.is_filled`, the `attrs` should be probably accessed via `component_vars.attrs` or `component_vars.fallthrough_attrs`.
3. Similarly to Vue, it should be the component class that decides if it wants to use the fallthrough attrs feature or not. I suggest an opt-in, so that, by default, user gets an error if they pass in unknown args/kwargs.
	```py
	class MyComp(component.Component):
		template_name = "abc.html"
		fallthrough_attrs = True
	```
	NOTE: While Vue sets this via `inheritAttrs`, I think `fallthrough_attrs` is more self-explanatory. (and googling for `fallthrough attrs` also gets you to the Vue docs)
4. This is the logic that extracts the "fallthrough attrs" from kwargs passed to `get_context_data()`
	```py
    argspec = inspect.getfullargspec(self.get_context_data)
	# NOTE: I use keyword-only args in my project, might want to add argspec.kwargs?
    known_prop_names = set([*argspec.args, *argspec.kwonlyargs])

    passthrough_props = {
        k: v for k, v in kwargs.items() if k not in known_prop_names
    }
	```
	So if I define `get_context_data` as:
	```py
	def get_context_data(self,  name, age=None, city=None):
		return {...}
	```
	then `known_prop_names` is `["name", "age", "city"]`

	and so if I render the component like so:
	```django
	{% component "my_comp" name="one" dogs_count=13 %}
	{% endcomponent %}
	```
	then `name` IS recognized, but `dogs_count` is NOT recognized. So `component_vars.fallthrough_attrs` gets populated with {"dogs_count": 13}
5. This bit from above:
	```py
	argspec = inspect.getfullargspec(self.get_context_data)
	known_prop_names = set([*argspec.args, *argspec.kwonlyargs])
	```
	could be done at class definition (`__new__`) so that we don't have to call `inspect.getfullargspec` on every render.

## Implementation
Implementation consists of 2 sub-features:
1. Allow to pass kwargs with special characters
2. Implementing the "fallthrough attrs" on the Component class

**Comment by tissieres** [comment]
For reference, [django-web-components](https://github.com/Xzya/django-web-components?tab=readme-ov-file#attributes-with-special-characters) supports these characters: `@:_-.` and also has a [nice way to handle attributes](https://github.com/Xzya/django-web-components?tab=readme-ov-file#passing-data-to-components) in general.

**Comment by EmilStenstrom** [comment]
I think this is a very valid use-case, passing things like `@click.stop="$dispatch('user_delete', { role })"` through a component.

I'm not sure about the "invalid attribute pass-through to attrs" solution though. I see that slippers and django-web-components have solved it this way. But I don't like that we're adding "magic" that transforms `@click.stop` to `attrs` (or `component_vars.attrs`) without it being explicit in the code. People reading the code will likely not understand it without going to the docs, which I'd like to avoid if possible.

An easier way to follow this would be if attrs was right there as an attribute:

```django
{% component "icon"
  icon="outline_trash"
  attrs="@click.stop=\"$dispatch('user_delete', { role })\""
%}{% endcomponent %}
```

(this would necessitate we don't mess up escaped quotes...)

... but I guess that limits the usability when you want to add more than one attribute. Maybe we could solve this with newlines?

```django
{% component "icon"
  icon="outline_trash"
  attrs="
      @click.stop=\"$dispatch('user_delete', { role })\"
      x-data=\"{ open: false }\"
  "
%}{% endcomponent %}
```
or with triple quotes?


```django
{% component "icon"
  icon="outline_trash"
  attrs="""
      @click.stop="$dispatch('user_delete', { role })"
      x-data="{ open: false }"
  """
%}{% endcomponent %}
```

**Reply by JuroOravec** [reply]
Going from what django-components has now (nothing) to the solution with `attrs` or `component_vars.attrs` skips one evolution of the API. In my project I too started with just explicit `attrs` prop like
```
attrs="@click.stop=\"$dispatch('user_delete', { role })\""
```
But I quickly found out that with already 2-3 attributes, it's difficult to read, because of all the quotes, and because
it doesn't wrap:
```
attrs="@click.stop=\"$dispatch('user_delete', { role })\" x-data='{ open: false }' class='pa-4 text-center'"
```

Your suggestion on using triple quotes in templates is interesting, and could address this.

However, there's another problem with a single `attrs` prop, which is that it's difficult to pass context variables to individual attributes.

E.g. if I want to dynamically set the `class` attribute, I can set it like any other attribute:
```django
{% component "icon"
  icon="outline_trash"
  @click.stop="$dispatch('user_delete', { role })"
  class=my_class_str
%}{% endcomponent %}
```

But if all attributes are rendered as string, I have to somehow insert the `my_class_str` variable into the `attrs` string (using the [add filter](https://docs.djangoproject.com/en/5.0/ref/templates/builtins/#add)), which again isn't that easy to read:
```django
{% component "icon"
  attrs="
	@click.stop=\"$dispatch('user_delete', { role })\"
	class=\""|add:my_class_str|add:"\"
  "
%}{% endcomponent %}
```

While you could also solve this by defining the `attrs` variable on Python side (inside `get_context_data`), you then have to define the `@click.stop` event handler on Python side too, which is disconnected from its use so it's not that great either.

So IMO from what I've seen, collecting the attributes into `attrs` or `component_vars.attrs` context variable is the cleanest solution I've seen so far.

---

However, I do agree that this behavior of collecting attributes in `component_vars.attrs` should be explicit. IMO this is an advanced feature - after all, I think it's most useful when creating UI library components*. It's not that relevant when writing business logic components. And I think that for advanced features, it's justified to require users to read the docs first.

\* My inspiration comes from Vue/Vuetify, where as a user of UI lib components, I can pass CSS classes or extra attrs to each component, which I use 90% of the time to tweak styling and event handlers.

Hence I think it should be opt-in. By default, something like
```django
{% component "icon"
  icon="outline_trash"
  @click.stop="$dispatch('user_delete', { role })"
  class=my_class_str
%}{% endcomponent %}
```

Should raise an error if `class` or `@click.stop` are NOT defined as inputs on the `get_context_data` method.

And only if I set some option on the component class, e.g.:
```py
class MyComp(component.Component):
	template_name = "abc.html"
	fallthrough_attrs = True
```

then the extra `class` and `@click.stop` would be accept and collected into `component_vars.attrs`.

**Reply by dylanjcastillo** [reply]
What about using something like this? It has some magic but makes it a bit more explicit as well.

```jinja
{% component "icon"
  icon="outline_trash"
  attrs.@click.stop="$dispatch('user_delete', { role })"
  attrs.class=my_class_str 
%} {% endcomponent %}
```

**Reply by JuroOravec** [reply]
@dylanjcastillo love that! Haven't thought of that before. 

I'd change the dot to colon or hash, as dot in attributes sometimes as the connotation of "modifier":
- colon (`attrs:`) - (my preference) colon is sometimes used for scoping values, like when you think of [AWS ARN](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference-arns.html) or [Redis keys](https://stackoverflow.com/questions/3554888/what-is-the-purpose-of-colons-within-redis-keys)
- hash (`attrs#`) - simply because it's not commonly used in attributes, so should be unique enough to catch attention

**Reply by dylanjcastillo** [reply]
I also like the colon version more!

**Reply by EmilStenstrom** [reply]
I like having attrs:param=x as an API because it's explicit, but I must say it doesn't look very elegant. 

Another option: What if we just started passing through variables to the get_context_data without the need for attrs at all?

```django
{% component "icon"
  icon="outline_trash"
  @click.stop="$dispatch('user_delete', { role })"
  x-data="{ open: false }"
%}{% endcomponent %}
```

Then in your get_context_data in attrs (or some other variable) you would do:
```python
def get_context_data(**kwargs):
    return {"attrs": self.kwargs_to_html_attr(kwargs, "@click.stop", "x-data")}
```

**Reply by tissieres** [reply]
Merging attributes is also quite useful, especially when using CSS utility classes. It could be specified as an option in the component class for example.

**Reply by JuroOravec** [reply]
@tissieres on the topic of merging attributes, what do you think of how it was done in [django-web-components](https://github.com/Xzya/django-web-components#default--merged-attributes)?

**Reply by tissieres** [reply]
I like having a way to specify that I want to merge some attributes with those already specified in the component template, but I don't like the fact that the `class` attribute is not handled as the other attributes in `django-web-components`. Using a template tag does not seem appropriate but I was not able to propose a better idea in this discussion.

**Comment by EmilStenstrom** [comment]
So, let summarize the different options:

A) Magic attrs (similar to Vue and django-web-components)
* `component.fallthrough_attrs = True`
* `{% component @click.stop="..." x-data="..." %}`
* `<div {{ attrs }}>`

B) Allow special symbols and handle everything in get_context_data
* `{% component @click.stop="..."  x-data="..." %}`
* `def get_context_data(**kwargs): return {"attrs": kwargs["@click.stop"] + " " + kwargs["x-data"]}`
* `<div {{ attrs }}>`

C) Namespaced parameters
* `{% component attrs:@click.stop="..." attrs:x-data="..."  %}`
* `def get_context_data(**attrs): return {"attrs": attrs}`
* `<div {{ attrs }}>`

D) Triple quoted parameters (no special chars in params needed)
* `{% component attrs="""@click.stop="..." \n x-data="..." """ %}`
* `def get_context_data(attrs): return {"attrs": attrs}`
* `<div {{ attrs }}>`

Did I miss any?

**Reply by JuroOravec** [reply]
Thanks for summarizing it! I'm not a fan of B (in it's current form). One nice feature that Vue's behavior and other options here have is that the user of a component can specify any attributes without the author of the component having to micro-manage it. So option B wouldn't play nicely into building component libraries.

**Reply by JuroOravec** [reply]
On option D - just brainstorming - maybe we could get away with using just a string even for variables?

E.g. so if I have this

```django
{% component "icon"
  icon="outline_trash"
  attrs="""
      @click.stop="$dispatch('user_delete', { role })"
      x-data="{ open: false }"
  """
%}{% endcomponent %}
```

and I want to add an attribute `class` with value from context variable `my_class`,
I could simply do?

```django
{% component "icon"
  icon="outline_trash"
  attrs="""
      @click.stop="$dispatch('user_delete', { role })"
      x-data="{ open: false }"
      class=my_class
  """
%}{% endcomponent %}
```

How that would have to work is that:
1. At parse time, we take the `attrs` string and chop it up into:
	- `@click.stop="$dispatch('user_delete', { role })"`
	- `x-data="{ open: false }"`
	- `class=my_class`
2. For each kwarg, we split it and find it's key and value:
	- key: `@click.stop`, value: `"$dispatch('user_delete', { role })"`
	- key: `x-data`, value: `"{ open: false }"`
	- key: `class`, value: `my_class`
3. And we pass this data to the ComponentNode.
4. Now, at render time, we treat the values of kwargs as "FilterExpression" (AKA context variable expression like `{{ my_var|some_filter }}`). It's similar to how we handle fill or slot names - that we can use a string literal like `{% slot "slot_name" %}`, or we can use a variable like `{% slot slot_name_var %}`.
5. When evaluated against the current Context, we get:
	- `"$dispatch('user_delete', { role })"` -> `"$dispatch('user_delete', { role })"`
	- `"{ open: false }"` -> `"{ open: false }"`
	- `my_class` -> the value of `context["my_class"]`

So this way, it would be enough for users just to use the triple-quotes!

Altho, so I just tried to define a component kwarg with triple-quotes, and well, it doesn't work 😅 So there's ~two~ one other issue with the triple-quotes:
1. We'd need to override the kwarg-parsing for components. This actually shouldn't be an issue after the addition of the `template_parser.py` file, as the function we'd need to override, `parse_bits`, is there.
2. ~Using the triple-quote attrs wouldn't look very pleasing for inlined templates:~
	```py
    class DashboardComponent(component.Component):
        template: types.django_html = """
			{% component "icon"
			  icon="outline_trash"
			  attrs=\"\"\"
			      @click.stop="$dispatch('user_delete', { role })"
			      x-data="{ open: false }"
			      class=my_class
			  \"\"\"
			%}{% endcomponent %}
        """
	```

	Altho actually we could always alternate between single and double quotes:
	```py
    class DashboardComponent(component.Component):
        template: types.django_html = """
			{% component "icon"
			  icon="outline_trash"
			  attrs='''
			      @click.stop="$dispatch('user_delete', { role })"
			      x-data="{ open: false }"
			      class=my_class
			  '''
			%}{% endcomponent %}
        """
	```

**Reply by JuroOravec** [reply]
Update: Managed to change the parsing regex to support triple-quotes for kwarg values, so something like this can be parsed:
```django
{% component "calendar" date='''2020-06-06''' %}
{% endcomponent %}
```

BUT, multi-line strings do not work out of the box:
```django
{% component "calendar"
	date='''
		2020-06-06
	'''
%}
{% endcomponent %}
```

Because Django's template parser does not handle multi-line tags by default. For that, the user (or us) would need to set [this workaround](https://stackoverflow.com/questions/49110044/django-template-tag-on-multiple-line/54206609#54206609):
```py
import re
from django.template import base
base.tag_re = re.compile(base.tag_re.pattern, re.DOTALL)
```

I think that for component tags it makes sense to be able to have them span multiple lines, so I can imagine that the workaround could be part of the installation step.

---

Alternatively, we would have to make a subclass of Django's Template, and override [`Template.compile_nodelist()`](https://github.com/django/django/blob/main/django/template/base.py#L175) and change the call to [`lexer.tokenize()`](https://github.com/django/django/blob/main/django/template/base.py#L187)

**Reply by JuroOravec** [reply]
So to summarize (sorry for spamming), if we want clear and explicit (non-magicky) interface, then I'd vote for either approach C) or D). D is imo slightly cleaner/explicit, but requires user to configure `base.tag_re`.

**Reply by EmilStenstrom** [reply]
I think just reading through your research about option D) above it seems that supporting triple quotes is WAY too much work and increased surface area. Thanks for establishing that! In my mind D) is out :)

I think I like C most even though it's a bit magic, but B is good too, and A is OK i guess :) What's your preference?

**Reply by JuroOravec** [reply]
I'd rule out B) too, because, [as mentioned here](https://github.com/EmilStenstrom/django-components/discussions/471#discussioncomment-9322215), it'd take away some flexibility from component libraries. Component library meaning a set of UI components used throughout a project, a [design system](https://en.wikipedia.org/wiki/Design_system#:~:text=A%20design%20system%20is%20a,as%20mobile%20applications%20or%20websites.).

**Reply by tissieres** [reply]
As a simple user, I like magic so my preference goes to A), followed by C). But C) can maybe be extended with some more magic for merging for example. Something like this `merge_attrs:x-data="..."`.

**Reply by dylanjcastillo** [reply]
I’d also vote for C

**Reply by EmilStenstrom** [reply]
@tissieres C) would include merging I think, so if you say `{% component "foobar" attrs:x=1 attrs:y=2 %}` you could render that with `<div {{ attrs }}>` and have with x=1 and y=2 included. Does that match what you would expect merge_attrs to do?

**Reply by EmilStenstrom** [reply]
Sounds like we have consensus around option C. I'll write up an issue.

**Reply by tissieres** [reply]
No, I was referring to the situation where the template already have an attribute with the same name and we want to merge the value given in the component with those. For example the template is `<div class="bg-red-100">` and the component `{% component "foobar" merge_attrs:class="text-xl" %}`, then instead of `<div class="text-xl">`, we would have `<div class="text-xl bg-red-100">`.

**Reply by JuroOravec** [reply]
In case of merging attributes, I think it makes most sense to think about this from the perspective of a component author (as opposed to component user), so I'm leaning towards taking a similar approach to [django-web-components](https://github.com/Xzya/django-web-components#default--merged-attributes). That is, merging would be done with a template tag that accepts a dict of our attributes, and any other extra attributes as defaults. `class` attributes are merged, while other attributes would be overriden.

The reason for this is that, as component author:
- I may NEED for a component to have some specific CSS class(es), maybe because I use it for styling, or there's some logic associated with it. So while I want to allow component users to supply their own CSS classes, I do NOT want to let them override mine:
	```html
	<div class="my_comp EXTERNAL_CLASSES"> <!-- I'm OK with adding component user's classes -->
	<div class="EXTERNAL_CLASSES"> <!-- I'm NOT OK with component user's classes overriding my internal ones -->
	```

	CSS classes are a bit of a special case, because many popular frameworks set the styling via them (e.g. Tailwind, Bootstrap, etc...). So if component user is trying to modify styling of the underlying HTML tag, there's a good chance they will use the `class` attribute for that.

	- We could take similar approach for the `style` attribute, as that may be used frequently too. But generally, a prefered approach is to move the styling into a CSS, so there's once again only the `class` attribute on the HTML tag that decides on the styling.

- For every other HTML attribute, the case is the opposite - there's a low chance that the user will want to override the HTML attributes that I've set. But if they do, they can do so, but it's their responsibility. Meaning that, as component author, I will supply my value for a certain attribute, but I do not think too hard about trying to merge or resolve the values. The HTML attribute API surface is huge, my time is limited, and it has minimal impact. So letting user override the values is a good enough solution.

For context, the perspective of component user is:
- I want to supply my own HTML attributes, so I can customize styling, event handling, or other behavior.

	- NOTE: As @tissieres pointed out there may be cases where they would want to forcefully override the `class` attribute instead of merging it, to achieve a specific styling. So they would achieve the latter case in:
		```html
		<div class="my_comp EXTERNAL_CLASSES">
		<div class="EXTERNAL_CLASSES">
		```

		However, as said earlier, this could break the component. So instead, I believe it should be up to the component author to make the component accept some kind of input that, when supplied, would render `<div class="EXTERNAL_CLASSES">` instead of `<div class="my_comp EXTERNAL_CLASSES">`

		So long story short, IMO component user should NOT use `merge_attrs:`. Instead, the component should be refactored. Ofc,  there may be a need for something similar for when component user does not have access to the component source code (e.g. 3rd party lib). But I would leave that for later iteration.

**Reply by EmilStenstrom** [reply]
See ticket here: https://github.com/EmilStenstrom/django-components/issues/488

**Reply by JuroOravec** [reply]
Awesome!

Btw, could I get a quick review on this one to clear it out of the way? https://github.com/EmilStenstrom/django-components/pull/482

UPDATE: Merged, thanks!

**Reply by JuroOravec** [reply]
Hey all, I made a first go at the feature at https://github.com/EmilStenstrom/django-components/pull/491 (WIP), please let me what you think!

**Comment by JuroOravec** [comment]
Hi all, this feature has just been released as v0.74 🎉 See the [README section "Rendering HTML attributes"](https://github.com/EmilStenstrom/django-components/tree/0.74?tab=readme-ov-file&rgh-link-date=2024-05-12T10%3A18%3A27Z#rendering-html-attributes)
